src/ott/neural/solvers/expectile_neural_dual.py

```diff
@@ -0,0 +1,655 @@
+
+__all__ = [
+    "BidirectionalPotentials", "ForwardPotentials", "EuclideanPotentials",
+    "DotCost", "PotentialModelWrapper", "PotentialTrainState", "MLP",
+    "ExpectileNeuralDual"
+]
+
+Train_t = Dict[Literal["train_logs", "valid_logs"], Dict[str, List[float]]]
+Callback_t = Callable[[int, DualPotentials], None]
+PotentialValueFn_t = Callable[[jnp.ndarray], jnp.ndarray]
+PotentialGradientFn_t = Callable[[jnp.ndarray], jnp.ndarray]
+CostFn_t = TypeVar("CostFn_t", bound=costs.CostFn)
+
+
+class BidirectionalPotentials(DualPotentials):
+  r"""The dual potentials for bidirectional mapping with TICost.
+
+  Args:
+  g: The second dual potential function.
+  grad_f: Gradient of the first dual potential function.
+  cost_fn: The translation invariant (TI) cost function used to solve
+  the OT problem.
+  """
+
+  def __init__(
+      self, g: PotentialValueFn_t, grad_f: PotentialGradientFn_t,
+      cost_fn: costs.TICost
+  ):
+
+    def g_cost_conjugate(x: jnp.ndarray) -> jnp.ndarray:
+      grad_h_inv = jax.grad(cost_fn.h_legendre)
+      y_hat = jax.lax.stop_gradient(x - grad_h_inv(grad_f(x)))
+      return -g(y_hat) + cost_fn(x, y_hat)
+
+    super().__init__(g_cost_conjugate, g, cost_fn=cost_fn, corr=False)
+    self.__grad_f = grad_f
+
+    assert isinstance(cost_fn, costs.TICost), (
+        "Cost must be a `TICost` and "
+        "provide access to Legendre transform of `h`."
+    )
+
+  @property
+  def _grad_f(self) -> Callable[[jnp.ndarray], jnp.ndarray]:
+    return jax.vmap(self.__grad_f)
+
+  def transport(self, vec: jnp.ndarray, forward: bool = True) -> jnp.ndarray:
+    """Transport from source to the target distribution or vice-versa."""
+    vec = jnp.atleast_2d(vec)
+    if forward:
+      return vec - self._grad_h_inv(self._grad_f(vec))
+    return vec - self._grad_h_inv(self._grad_g(vec))
+
+
+class ForwardPotentials(DualPotentials):
+  r"""The dual potentials for only forward mapping.
+
+  Args:
+  g: The second dual potential function.
+  grad_f: Gradient of the first dual potential function.
+  cost_fn: The cost function used to solve the OT problem.
+  """
+
+  def __init__(
+      self, g: PotentialValueFn_t, grad_f: PotentialGradientFn_t,
+      cost_fn: costs.CostFn
+  ):
+
+    def g_cost_conjugate(x: jnp.ndarray) -> jnp.ndarray:
+      y_hat = jax.lax.stop_gradient(grad_f(x))
+      return -g(y_hat) + cost_fn(x, y_hat)
+
+    super().__init__(g_cost_conjugate, g, cost_fn=cost_fn, corr=False)
+    self.__grad_f = grad_f
+
+  @property
+  def _grad_f(self) -> Callable[[jnp.ndarray], jnp.ndarray]:
+    return jax.vmap(self.__grad_f)
+
+  def transport(self, source: jnp.ndarray, forward: bool = True) -> jnp.ndarray:
+    """Transport the points from source to the target distribution."""
+    source = jnp.atleast_2d(source)
+    assert (forward
+            is True), ("Only forward mapping (source -> target) is supported.")
+    return self._grad_f(source)
+
+
+class EuclideanPotentials(DualPotentials):
```

Comment:
I think we have a varriable `corr` in the original potentials class to enable this slightly different logic, would it work here? (at least the `transport` method would not need to be changed)

Reply:
I agree, overriding methods was unnecessary in this class and we may use `corr` instead. I just suggested that it would be better to use inheritance instead of `corr`  in `DualPotentials`.

Reply:
Class EuclideanPotentials removed.